File: source/includes/aggregation/multi-field-join.py
const { MongoClient } = require("mongodb");

const uri = "<connection string>";
const client = new MongoClient(uri);

async function run() {
  try {
    const aggDB = client.db("agg_tutorials_db");

    // start-colls
    const productsColl = await aggDB.collection("products");
    const ordersColl = await aggDB.collection("orders");
    // end-colls

    // start-insert-products
    await productsColl.deleteMany({});

    const productsData = [
      {
        name: "Asus Laptop",
        variation: "Ultra HD",
        category: "ELECTRONICS",
        description: "Great for watching movies",
      },
      {
        name: "Asus Laptop",
        variation: "Standard Display",
        category: "ELECTRONICS",
        description: "Good value laptop for students",
      },
      {
        name: "The Day Of The Triffids",
        variation: "1st Edition",
        category: "BOOKS",
        description: "Classic post-apocalyptic novel",
      },
      {
        name: "The Day Of The Triffids",
        variation: "2nd Edition",
        category: "BOOKS",
        description: "Classic post-apocalyptic novel",
      },
      {
        name: "Morphy Richards Food Mixer",
        variation: "Deluxe",
        category: "KITCHENWARE",
        description: "Luxury mixer turning good cakes into great",
      },
    ];

    await productsColl.insertMany(productsData);
    // end-insert-products

    // start-insert-orders
    await ordersColl.deleteMany({});

    const orderData = [
      {
        customer_id: "elise_smith@myemail.com",
        orderdate: new Date("2020-05-30T08:35:52Z"),
        product_name: "Asus Laptop",
        product_variation: "Standard Display",
        value: 431.43,
      },
      {
        customer_id: "tj@wheresmyemail.com",
        orderdate: new Date("2019-05-28T19:13:32Z"),
        product_name: "The Day Of The Triffids",
        product_variation: "2nd Edition",
        value: 5.01,
      },
      {
        customer_id: "oranieri@warmmail.com",
        orderdate: new Date("2020-01-01T08:25:37Z"),
        product_name: "Morphy Richards Food Mixer",
        product_variation: "Deluxe",
        value: 63.13,
      },
      {
        customer_id: "jjones@tepidmail.com",
        orderdate: new Date("2020-12-26T08:55:46Z"),
        product_name: "Asus Laptop",
        product_variation: "Standard Display",
        value: 429.65,
      },
    ];

    await ordersColl.insertMany(orderData);
    // end-insert-orders

    const pipeline = [];

    // start-embedded-pl-match1
    const embedded_pl = [];

    embedded_pl.push({
      $match: {
        $expr: {
          $and: [
            { $eq: ["$product_name", "$$prdname"] },
            { $eq: ["$product_variation", "$$prdvartn"] },
          ],
        },
      },
    });
    // end-embedded-pl-match1

    // start-embedded-pl-match2
    embedded_pl.push({
      $match: {
        orderdate: {
          $gte: new Date("2020-01-01T00:00:00Z"),
          $lt: new Date("2021-01-01T00:00:00Z"),
        },
      },
    });
    // end-embedded-pl-match2

    // start-embedded-pl-unset
    embedded_pl.push({
      $unset: ["_id", "product_name", "product_variation"],
    });
    // end-embedded-pl-unset

    // start-lookup
    pipeline.push({
      $lookup: {
        from: "orders",
        let: {
          prdname: "$name",
          prdvartn: "$variation",
        },
        pipeline: embedded_pl,
        as: "orders",
      },
    });
    // end-lookup

    // start-match
    pipeline.push({
      $match: {
        orders: { $ne: [] },
      },
    });
    // end-match

    // start-unset
    pipeline.push({
      $unset: ["_id", "description"],
    });
    // end-unset

    // start-run-agg
    const aggregationResult = await productsColl.aggregate(pipeline);
    // end-run-agg

    for await (const document of aggregationResult) {
      console.log(document);
    }
  } finally {
    await client.close();
  }
}

run().catch(console.dir);

File: content/manual/v8.2/source/includes/aggregation/aggregation-examples/multi-field-join/full-files/multi-field-join.rs
use mongodb::{
    bson::{doc, DateTime, Bson, Document},
    Client,
    Collection,
};
use serde::{Deserialize, Serialize};
use futures::stream::TryStreamExt;
use std::error::Error;

// start-structs
#[derive(Debug, Serialize, Deserialize)]
struct Product {
    name: String,
    variation: String,
    category: String,
    description: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct Order {
    customer_id: String,
    order_date: DateTime,
    product_name: String,
    product_variation: String,
    value: f32,
}
// end-structs

#[tokio::main]
async fn main() -> mongodb::error::Result<()> {
    let uri = "<connection string>";
    let client = Client::with_uri_str(uri).await?;
    let agg_db = client.database("agg_tutorials_db");

    // start-insert-sample-data
    let products: Collection<Product> = agg_db.collection("products");
    let orders: Collection<Order> = agg_db.collection("orders");

    products.delete_many(doc! {}).await?;
    orders.delete_many(doc! {}).await?;

    let product_docs = vec![
        Product {
            name: "Asus Laptop".to_string(),
            variation: "Ultra HD".to_string(),
            category: "ELECTRONICS".to_string(),
            description: "Great for watching movies".to_string(),
        },
        Product {
            name: "Asus Laptop".to_string(),
            variation: "Standard Display".to_string(),
            category: "ELECTRONICS".to_string(),
            description: "Good value laptop for students".to_string(),
        },
        Product {
            name: "The Day Of The Triffids".to_string(),
            variation: "1st Edition".to_string(),
            category: "BOOKS".to_string(),
            description: "Classic post-apocalyptic novel".to_string(),
        },
        Product {
            name: "The Day Of The Triffids".to_string(),
            variation: "2nd Edition".to_string(),
            category: "BOOKS".to_string(),
            description: "Classic post-apocalyptic novel".to_string(),
        },
        Product {
            name: "Morphy Richards Food Mixer".to_string(),
            variation: "Deluxe".to_string(),
            category: "KITCHENWARE".to_string(),
            description: "Luxury mixer turning good cakes into great".to_string(),
        },
    ];
    products.insert_many(product_docs).await?;

    let order_docs = vec![
        Order {
            customer_id: "elise_smith@myemail.com".to_string(),
            order_date: DateTime::builder().year(2020).month(5).day(30).hour(8).minute(35).second(52).build().unwrap(),
            product_name: "Asus Laptop".to_string(),
            product_variation: "Standard Display".to_string(),
            value: 431.43,
        },
        Order {
            customer_id: "tj@wheresmyemail.com".to_string(),
            order_date: DateTime::builder().year(2019).month(5).day(28).hour(19).minute(13).second(32).build().unwrap(),
            product_name: "The Day Of The Triffids".to_string(),
            product_variation: "2nd Edition".to_string(),
            value: 5.01,
        },
        Order {
            customer_id: "oranieri@warmmail.com".to_string(),
            order_date: DateTime::builder().year(2020).month(1).day(1).hour(8).minute(25).second(37).build().unwrap(),
            product_name: "Morphy Richards Food Mixer".to_string(),
            product_variation: "Deluxe".to_string(),
            value: 63.13,
        },
        Order {
            customer_id: "jjones@tepidmail.com".to_string(),
            order_date: DateTime::builder().year(2020).month(12).day(26).hour(8).minute(55).second(46).build().unwrap(),
            product_name: "Asus Laptop".to_string(),
            product_variation: "Standard Display".to_string(),
            value: 429.65,
        },
    ];
    orders.insert_many(order_docs).await?;
    // end-insert-sample-data

    let mut pipeline = Vec::new();

    // start-embedded-pl-match1
    let mut embedded_pipeline = Vec::new();
    embedded_pipeline.push(doc! {
        "$match": {
            "$expr": {
                "$and": [
                    { "$eq": ["$product_name", "$$prdname"] },
                    { "$eq": ["$product_variation", "$$prdvartn"] }
                ]
            }
        }
    });
    // end-embedded-pl-match1

    // start-embedded-pl-match2
    embedded_pipeline.push(doc! {
        "$match": {
            "order_date": {
                "$gte": DateTime::builder().year(2020).month(1).day(1).build().unwrap(),
                "$lt": DateTime::builder().year(2021).month(1).day(1).build().unwrap()
            }
        }
    });
    // end-embedded-pl-match2

    // start-embedded-pl-unset
    embedded_pipeline.push(doc! {
        "$unset": ["_id", "product_name", "product_variation"]
    });
    // end-embedded-pl-unset

    // start-lookup
    pipeline.push(doc! {
        "$lookup": {
            "from": "orders",
            "let": {
                "prdname": "$name",
                "prdvartn": "$variation"
            },
            "pipeline": embedded_pipeline,
            "as": "orders"
        }
    });
    // end-lookup

    // start-match
    pipeline.push(doc! {
        "$match": {
            "orders": { "$ne": [] }
        }
    });
    // end-match

    // start-unset
    pipeline.push(doc! {
        "$unset": ["_id", "description"]
    });
    // end-unset

    // start-run-agg
    let mut cursor = products.aggregate(pipeline).await?;
    // end-run-agg

    while let Some(result) = cursor.try_next().await? {
        println!("* {:?}", result);
    }

    Ok(())
}
